feat(gatsby-source-graphql): Add transformSchema option

ariadne-github opened the pull request:
## Description

Should resolve #23552. Adds an optional custom transformSchema option to gatsby-source-graphql that passes the default transforms, schema and link. I has many potential usages, but the consumer should be able to use the same default transforms or tweak or replace them completely. The original use case is to add an argument to the generated "fieldName" field that can then be read by a custom apollo-link to customize the underlying http call with a particular header. Tagging @vladar. This is a proposed implementation that should be safe when not used, and works for my use case. I'm not sure if other parameters might be passed along the transformSchema arguments, maybe plugin options? Maybe the graphql-tools default transformSchema?

### Documentation

Example usage that replaces the third default transform and the default link.
```javascript
// gatsby-config
const { transformSchema } = require(`graphql-tools`);

function customTransformSchema({ schema, link, resolver, defaultTransforms }) {
	const transforms = [defaultTransforms[0], defaultTransforms[1]];
	return transformSchema(
		{
			schema,
			link: myCustomLink(link),
		 },
		[
			defaultTransforms[0], 
			defaultTransforms[1], 
			new MyCustomNamespaceUnderFieldTransform(resolver)
		]
    )
};

{
	...
	plugins: [
		...
		{
		  resolve: "gatsby-source-graphql",
		  options: {
			typeName: "myTypeName",
			fieldName: "myField",
			url: "myEndpoint",
			transformSchema: customTransformSchema
		  },
		},
		...
	],
	...
}



```
@gatsbyjs/learning.

## Related Issues

Resolves #23552.

yaacovCR commented:
Anything helpful we can do from the graphql-tools end? Note that graphql-tools WrapFields/WrapType in general might be useful...

ariadne-github commented:
@vladar I merged again the most recent commits from master and added a section in the README.md. The code is taken from a working example. I have only my use case, which is a bit convoluted. Does anyone see a better approach?

vladar commented on packages/gatsby-source-graphql/README.md:
```diff
@@ -150,6 +150,153 @@ module.exports = {
+- resolver (default resolver)
+- defaultTransforms (an array with the default transforms)
+- options (plugin options)
+
```
I think we need to add a note that the return value is expected to be the final schema used for stitching.

ariadne-github replied:
So, basically, I should only leave this as a reference?

```js
// transform-schema.js
const { wrapSchema } = require(`@graphql-tools/wrap`)
const { linkToExecutor } = require(`@graphql-tools/links`)

const transformSchema = ({
  schema,
  link,
  resolver,
  defaultTransforms,
  options,
}) => {
  return wrapSchema(
    {
      schema,
      executor: linkToExecutor(link),
    },
    defaultTransforms
  )
}

module.exports = transformSchema
```
I just think that would be a bit vague for anyone else to do something useful without knowing at least how the `NamespaceUnderFieldTransform` is implemented. But if it's good for you, it's good for me too.

What if I link to this issue somewhere to leave a hint?

vladar replied:
The problem with this feature in general (as I see it) is that it requires a deep understanding of plugin internals and `graphql-tools` anyways.

I doubt that the example will be enough for anyone to actually use it. People will probably have to dig into the plugin source code anyways. But it may be rather confusing to other people who don't use this feature.

So we need to surface that this feature exists in docs and leave hints on what other knowledge people may need to actually use it.

Adding a link to the original issue makes sense to me though. Also, let's wait for the learning team review.

yaacovCR commented:
To update my comment above, you could keep as is, just rename to customWrapSchemaFn, and then graphql-tools wrapSchema function could be replaced entirely with a custom implementation.

My unsolicited two cents is that transformSchema a bit too generic...

ariadne-github commented:
> To update my comment above, you could keep as is, just rename to customWrapSchemaFn, and then graphql-tools wrapSchema function could be replaced entirely with a custom implementation.
> 
> My unsolicited two cents is that transformSchema a bit too generic...

I've updated the code as requested by @yaacovCR, and I also updated and simplified the documentation.

yaacovCR commented:
createWrappedSchema?

vladar commented:
I think `createWrappedSchema` is easy to confuse with `createSchema`. I actually think we should revert back to `transformSchema`.

I understand that for someone closely familiar with `graphql-tools` this sounds ambiguous but for the plugin user it clearly communicates the intent of this option. If you are ready to dive into `graphql-tools` level then I don't think this ambiguity will confuse you.

yaacovCR commented:
Hi!

I agree with you, but for a different reason. From gatsby perspective, calling wrapSchema is not the only valid way of transforming the original schema, so transformSchema is better.

yaacovCR commented:
Sorry @ariadne-github for throwing things off track!

ariadne-github commented:
In conclusion, do you agree on **transformSchema**?

yaacovCR commented:
Yes. In general, I have found myself to essentially always agree with @vladar :)
